### ISSUE TITLE
PHPUnit output is not visible until the tests are done (buffering problem?)

### ISSUE THREAD

**Issue by borekb** [mikeerickson/gulp-phpunit#17]
When I run PHPUnit on command-line I see dots and other markers as the tests progress but with gulp-phpunit, there is no output at all until all tests are done.

Is there an option to disable output buffering?

**Comment by mikeerickson** [comment]
@borekb This has been added to 0.8.0, will be live shortly

**Comment by WitteStier** [comment]
This bug is still alive for 0.9.1

[22:46:10] Using gulpfile /bm/gulpfile.js
[22:46:10] Starting 'test-server'...
[22:46:13] Finished 'test-server' after 2.8 s
[22:46:44] PHPUnit 4.7.7 by Sebastian Bergmann and contributors.

FF......S..........S...................F......................... 65 / 68 ( 95%)
...

Time: 17.44 seconds, Memory: 23.75Mb

**Comment by mikeerickson** [comment]
Actually, I believe this is a regression introduced in 0.9.x.

Can your trying using 

$ npm install gulp-phpunit@0.8.4

And see if it fixes the issue (this is the previous release before some refactoring in 0.9.x

**Comment by WitteStier** [comment]
$ npm uninstall gulp-phpunit
$ npm install gulp-phpunit@0.8.4
Now it prints the phpunit output as expected but with a little surprise on the end.

[21:11:47] Using gulpfile /bm/gulpfile.js
[21:11:47] Starting 'test-server'...
PHPUnit 4.7.7 by Sebastian Bergmann and contributors.
....
Generating code coverage report in HTML format ... done
[21:12:16] 'test-server' errored after 29 s
[21:12:16] Error: Command failed: /bin/sh -c ../vendor/bin/phpunit --colors -c "/bm/private/test/phpunit.xml"

    at ChildProcess.exithandler (child_process.js:751:12)
    at ChildProcess.emit (events.js:110:17)
    at maybeClose (child_process.js:1015:16)
    at Process.ChildProcess._handle.onexit (child_process.js:1087:5)

I guess this is the reason why the bug appears again after refactoring.
https://github.com/mikeerickson/gulp-phpunit/blob/master/test/test.js

**Comment by mikeerickson** [comment]
@WitteStier meaning the fact that I dont test for it is the reason for refactoring.    

FWIW, testing gulp plugins of this fashion is not an easy one.   

At any rate, thanks for confirming, I will get this cured into 0.9.2 this weekend

**Comment by mikeerickson** [comment]
@WitteStier So I have tweaked things, this is now appear (using gulp phpunit)


http://d.pr/i/Ajc6

**Comment by WitteStier** [comment]
PHPUnit is finished after 8.42ms? and i see the test results 2 times or do i overlook something?
at least the notify is new and helpful :)

**Comment by mikeerickson** [comment]
Time is slow as it is running on a vagrant machine (not locally)

**Comment by mikeerickson** [comment]
Hey, better to see something twice than not at all ;).  I have since cured that issue ...

**Comment by WitteStier** [comment]
I am looking forward to your commit. Maybe you can re-open this issue so its clear that your working on it. At least that it will not reported again.

**Comment by mikeerickson** [comment]
I didnt have time over the weekend to get this wrapped up.  Will try and get this merged into update in next couple days (and yes, reopening this ticket)

**Comment by jeffwa** [comment]
I encountered this issue today after updating to 0.9.1, though in my case PHPUnit wasn't even being run. The gulp task would run and there'd be output stating is was running the task, however, the task would output the following:

[10:55:02] Using gulpfile /path/to/gulpfile.js
[10:55:02] Starting 'phplint'...
[10:55:02] Finished 'phplint' after 527 ms
[10:55:02] Starting 'phpunit'...
[10:55:03] gulp-notify: [Gulp Task: PHPUnit] Task complete
[10:55:03] Finished 'phpunit' after 469 ms

indicating that it wasn't actually running the unit tests. Running my unit tests directly with PHPUnit takes ~6.8 seconds. (Using gulp I output coverage files which makes the full test suite take ~2min to complete.) Rolling back to 0.8.4 as previously suggested fixes the issue for me.

Just thought I'd add this in case someone else encounters the issue of unit tests not running and thinking this isn't related. (I also see the double output of PHPUnit statistics, but that's not really a concern as long as the tests run.)

**Comment by mikeerickson** [comment]
Addressed issues which would lead to no output in version 0.10.0

**Comment by jeffwa** [comment]
Finally had a chance to update from 0.8.4 to 0.11.2 (actually ended up just deleting my node_modules directory and starting from scratch) and I'm having the same issue as before:
PHPUnit 5.1.0 by Sebastian Bergmann and contributors.
Runtime:       PHP 5.6.16 with Xdebug 2.3.3

[15:30:05] Using gulpfile /path/to/gulpfile.js
[15:30:05] Starting 'phplint'...
[15:30:05] Finished 'phplint' after 480 ms
[15:30:05] Starting 'phpunit'...
[15:30:06] gulp-notify: [Gulp Task: PHPUnit] Task complete
[15:30:06] Finished 'phpunit' after 555 ms

As before, it's not running my unit tests. Running PHPUnit directly (without code coverage generation: 6.2 seconds, with code coverage generation: 1.6 minutes) takes longer than what is running. 

Looking at the commit history it looks like there's a major refactor underway* (or there was recently) and could have caused a regression. Is there a 'safe' version to pin to (other than 0.8.4) for the time being?

\* The last commit to the project on December 20 is unusual. All references to phpunit are changed to phplint. Is this a mistake?

**Comment by mikeerickson** [comment]
@jeffwa Yes, there was a rather significant refactor at 0.11 and while a regression could have surely manifested, I am not see the same thing you are reporting.

Further, with regards to phplint reference, yes indeed a comment mistake (I was working on phplint and phpunit at the same time).

My apologies for the regression on your issue, going to have to try and reproduce this locally as everything is working correctly on my end.   All tests are definitely executed (I use this plugin daily at my day job)

**Comment by mikeerickson** [comment]
@jeffwa While I have you, can you please share with me your task details and phpunit.xml file

**Comment by jeffwa** [comment]
Thanks for the quick follow up, @mikeerickson.

My task (stripped down to the basics):

    gulp.task('phpunit', function() {
        return gulp.src(__dirname + '/module/HelpAdmin/build/phpunit.xml')
            .pipe(plugins.phpunit('/usr/local/bin/phpunit'));
    });

phpunit.xml:

    <phpunit backupGlobals="false"
             bootstrap="../test/Bootstrap.php"
             beStrictAboutOutputDuringTests="true"
             beStrictAboutTestsThatDoNotTestAnything="true"
             beStrictAboutTodoAnnotatedTests="true"
             checkForUnintentionallyCoveredCode="true"
             verbose="true"
             colors="true">
        <testsuites>
            <testsuite name="controller">
                <directory suffix="Test.php">../test/HelpAdminTest/Controller</directory>
            </testsuite>
            <testsuite name="entity">
                <directory suffix="Test.php">../test/HelpAdminTest/Entity</directory>
            </testsuite>
            <testsuite name="model">
                <directory suffix="Test.php">../test/HelpAdminTest/Model</directory>
            </testsuite>
            <testsuite name="service">
                <directory suffix="Test.php">../test/HelpAdminTest/Service</directory>
            </testsuite>
        </testsuites>

        <logging>
            <log type="coverage-html" target="coverage"/>
        </logging>

        <filter>
            <blacklist>
                <directory>../../../vendor</directory>
            </blacklist>

            <whitelist processUncoveredFilesFromWhitelist="true">
                <directory suffix='.php'>../src/HelpAdmin</directory>
                <exclude>
                    <directory suffix='.php'>../src/HelpAdmin/Navigation</directory>
                    <directory suffix='.php'>../src/HelpAdmin/Annotation</directory>
                    <file>../src/HelpAdmin/Helpers/UploadHandler.php</file>
                </exclude>
            </whitelist>
        </filter>
    </phpunit>

**Comment by jeffwa** [comment]
I added some debug statements to index.js just to see if I could narrow down where the issue could be and it has something to do with the `exec(cmd, function (error, stdout, stderr) {` statement. (line 212 from 0.11.0 tagged version.) I added `console.log(cmd);` immediately before that line and it output:

    /usr/local/bin/phpunit --colors=always -c "/Users/bjmc/Development/src/web/help.bisnode.be/module/HelpAdmin/build/phpunit.xml"

When I copy that and run it directly in the terminal it executes `phpunit` as expected. Adding `console.log(cmd)` inside the `exec` callback is never executed.

**Comment by jeffwa** [comment]
After some more digging, the change from 

    process.stdout.write(str);

to 

    cb(null, str);

in commit https://github.com/mikeerickson/gulp-phpunit/commit/e64a34fad11019e5377af1bedac9a1e1d1a8c8ec is where the problem lies (at least in my case.) Returning output to `cb()` seems to trigger the finish of the calling task. When I revert the line in `index.js` back to `process.stdout.write(str);` it works fine.

**Comment by mikeerickson** [comment]
Thx for info, I will have a look at this over the weekend and get a fix released

**Comment by mikeerickson** [comment]
@jeffwa OK, can you please adjust your code as follows.  I have made this simple adjustment to a fresh install of Laravel 5.2 using PHPUnit 0.11.2 (version installed by nom install)

Note: I am using the default phpunit.xml file from default install, for my local testing.

````

var configFile  = __dirname + '/module/HelpAdmin/build/phpunit.xml';

var options = {
    configurationFile: configFile
    notify: false
};

elixir(function(mix) {
    mix.phpUnit('',options);
});
````

**Comment by mikeerickson** [comment]
@jeffwa Here is a screenshot of what I am seeing on my end using the above code

http://d.pr/i/3TuP

**Comment by jeffwa** [comment]
@mikeerickson Sorry for the delayed response. The code that is affected by this issue is a work project so I don't have access to it on the weekends. 

I'm not using Laravel so I can't use the code you supplied. I updated my call to gulp-phpunit to look like this:

    return gulp.src('')
        .pipe(plugins.phpunit('', {configurationFile: _phpunitXml, notify: false}));

(gulp complained if src was null) and still skips the tests.

I also tried calling it this way:

    return plugins.phpunit('', {configurationFile: _phpunitXml, notify: false});

without `gulp.src()` but it had the same results.

**Comment by mikeerickson** [comment]
Hey Jeff, my apologies for assuming you were using with Laravel.  I am out sick today so I don't have access to code at this time. 

I will try and have a look at this later today. I will do all j can to help you get this resolved today

**Comment by mikeerickson** [comment]
@jeffwa Please have a look at this quick sample app (uses Slim3 framework, but not a requirement)

https://github.com/mikeerickson/gulp-phpunit-slim3-example

The core of this is the gulp task and custom.phpunit.xml file

````
	var opts = {
		debug: false  // this is just here to make sure the proper command is executed
	};

	gulp.task('phpunit', function() {
		return gulp.src('custom.phpunit.xml')
			.pipe(phpunit('', opts));
	});
````

**Comment by jeffwa** [comment]
@mikeerickson No worries. And really, don't sweat this too much. I'm able to work around it so it's not a huge problem for me. (Also, I'm out sick today so I can't do anything about this until I'm back in the office.)

During my investigation of this issue I've actually edited my gulp task (the one on my work project that fails) to be identical to what you posted in your last message with no luck. It's entirely possible it's something in my setup.

That said, I cloned the sample app on my home setup, installed everything and ran `gulp phpunit` and it works fine. This is leading me to believe that my project is the issue. When I get to the office tomorrow (13 Jan) I'll try cloning the sample project there to see what happens and report back.

Thanks again for your work on this issue and the plugin itself. It's really useful.

**Comment by jeffwa** [comment]
@mikeerickson Cloning the sample project in my work dev environment, installing everything and running `gulp phpunit` works fine. 

I still have no idea what is causing the issue in my project that is making this not work. I've deleted `node_modules` and re-installed, tried wrapping my `gulpfile.js` functions in an IIFE, updated the configuration of `phpunit.xml` and re-written my `gulp.task` several different ways, all to no avail.

**Comment by mikeerickson** [comment]
@jeffwa hard to debug issues like this, but still prepared to offer whatever assistance I can,  how about your run the task with debug flag enabled to see what command is executed.  Further, can you give me a directory tree of your setup? 

$ tree | pbcopy 

  (making assumption you are on Mac). Otherwise expand your file structure enough to show tests directory and location of configuration file

**Comment by jeffwa** [comment]
@mikeerickson Yeah, it's been pretty frustrating on this end that's for sure. :) Especially since it works in that sample project in the same dev environment.

debug: [Version] 0.11.3 Debug Cmd: /usr/local/bin/phpunit --colors=always --debug -c "/Users/bjmc/Development/src/web/help.bisnode.be/module/HelpAdmin/build/phpunit.xml"

tree: (I removed the node_modules and vendor directories as the output was > 12k lines.

.
├── BANNER
├── README.md
├── build.xml
├── composer.json
├── composer.lock
├── config
│   ├── application.config.php
│   └── autoload
│       ├── development.php
│       ├── global.php
│       ├── local.php
│       ├── production.php
│       └── testing.php
├── gulpfile.js
├── init_autoloader.php
├── module
│   └── HelpAdmin
│       ├── Module.php
│       ├── build
│       │   └── phpunit.xml
│       ├── config
│       │   ├── design.development.php
│       │   ├── design.local.php
│       │   ├── design.production.php
│       │   ├── design.testing.php
│       │   ├── footer.config.php
│       │   └── module.config.php
│       ├── language
│       │   └── application
│       │       ├── en.php
│       │       ├── fr.php
│       │       └── nl.php
│       ├── src
│       │   └── HelpAdmin
│       │       ├── Annotation
│       │       │   └── AbstractAnnotation.php
│       │       ├── Controller
│       │       │   ├── AbstractBase.php
│       │       │   ├── Application.php
│       │       │   ├── Article.php
│       │       │   ├── Asset.php
│       │       │   ├── Deploy.php
│       │       │   ├── Help.php
│       │       │   ├── Index.php
│       │       │   └── Logout.php
│       │       ├── Entity
│       │       │   ├── AbstractBase.php
│       │       │   ├── Application.php
│       │       │   └── Article.php
│       │       ├── Helpers
│       │       │   ├── Rights.php
│       │       │   ├── RightsGroups.php
│       │       │   ├── SsoHelper.php
│       │       │   └── UploadHandler.php
│       │       ├── Model
│       │       │   ├── AbstractBase.php
│       │       │   ├── Application.php
│       │       │   └── Article.php
│       │       ├── Navigation
│       │       │   └── Service
│       │       │       └── AdminNavigationFactory.php
│       │       └── Service
│       │           ├── AbstractBase.php
│       │           ├── Application.php
│       │           ├── Article.php
│       │           └── Asset.php
│       ├── test
│       │   ├── Bootstrap.php
│       │   ├── HelpAdminTest
│       │   │   ├── Controller
│       │   │   │   ├── ApplicationTest.php
│       │   │   │   ├── ArticleTest.php
│       │   │   │   ├── AssetTest.php
│       │   │   │   ├── DeployTest.php
│       │   │   │   ├── IndexTest.php
│       │   │   │   └── LogoutTest.php
│       │   │   ├── Entity
│       │   │   │   ├── ApplicationTest.php
│       │   │   │   └── ArticleTest.php
│       │   │   ├── Model
│       │   │   │   ├── ApplicationTest.php
│       │   │   │   └── ArticleTest.php
│       │   │   └── Service
│       │   │       ├── ApplicationTest.php
│       │   │       └── ArticleTest.php
│       │   └── TestConfig.php
│       └── view
│           ├── applications
│           │   ├── index.phtml
│           │   ├── list.phtml
│           │   └── listItem.phtml
│           ├── articles
│           │   ├── detail.phtml
│           │   ├── index.phtml
│           │   ├── list.phtml
│           │   └── listItem.phtml
│           ├── assets
│           │   └── index.phtml
│           ├── common
│           │   ├── form.phtml
│           │   └── noItems.phtml
│           ├── deploy
│           │   └── index.phtml
│           ├── error
│           │   ├── 404.phtml
│           │   └── index.phtml
│           ├── json
│           │   └── json.phtml
│           └── layout
│               ├── blank.phtml
│               ├── flashMessages.phtml
│               ├── flashMessagesBlock.phtml
│               ├── layout.phtml
│               └── version.txt
├── package.json
├── public
│   ├── css
│   │   ├── help-admin.css
│   │   └── help-admin.min.css
│   ├── favicon.ico
│   ├── favicon.png
│   ├── index.php
│   ├── js
│   │   ├── blueimp
│   │   │   ├── javascript-canvas-to-blob
│   │   │   │   └── canvas-to-blob.min.js
│   │   │   ├── javascript-load-image
│   │   │   │   └── load-image.all.min.js
│   │   │   ├── javascript-templates
│   │   │   │   └── tmpl.min.js
│   │   │   └── jquery-file-upload
│   │   │       ├── jquery.fileupload-audio.js
│   │   │       ├── jquery.fileupload-image.js
│   │   │       ├── jquery.fileupload-jquery-ui.js
│   │   │       ├── jquery.fileupload-process.js
│   │   │       ├── jquery.fileupload-ui.js
│   │   │       ├── jquery.fileupload-validate.js
│   │   │       ├── jquery.fileupload-video.js
│   │   │       ├── jquery.fileupload.js
│   │   │       ├── jquery.iframe-transport.js
│   │   │       └── jquery.ui.widget.js
│   │   ├── help-admin.js
│   │   ├── help-admin.min.js
│   │   └── jquery.are-you-sure.js
│   └── src
│       ├── js
│       │   ├── main.js
│       │   └── sortable.js
│       └── scss
│           ├── ha-articles.scss
│           ├── ha-forms.scss
│           ├── ha-miscellaneous.scss
│           ├── ha-overrides.scss
│           └── help-admin.scss
└── vendor [...snipped...]

1936 directories, 11220 files

**Comment by mikeerickson** [comment]
Awesome, I will figure it out... Pretty bizarre that's for sure.  I assume we won't be able to converse about this again until you return to work Monday.  Where are you located ?

**Comment by mikeerickson** [comment]
One question, if you take the cmd that is displayed in debug output and execute from command line directly, does it work as expected?

**Comment by mikeerickson** [comment]
Additionally, if you do have command line execution working, can you please share what you are executing

**Comment by mikeerickson** [comment]
One more item, you can try defining the path to xml file as:

./modiule/build/phpunit.xml

**Comment by jeffwa** [comment]
@mikeerickson I'm in Brussels, Belgium. I'm to work today.

Running the command that's displayed in debug mode works fine. When I run it myself it works fine, too:

    # cd /Users/bjmc/Development/src/web/help.bisnode.be
    # phpunit -c ./module/HelpAdmin/build/phpunit.xml

Though usually if I run it manually I cd to the location of the phpunit.xml file and just run `phpunit`.

That path doesn't exist. The layout of the tree output above isn't aligned quite right. the correct path to phpunit.xml is:
    
    ./module/HelpAdmin/build/phpunit.xml

Running the task with that as the path still fails.

**Comment by mikeerickson** [comment]
Off to bed, one more idea for evening.  Put the phpunit.xml file at same level as gulpfile and then point to "phpunit.xml". I will pick this up again tomorrow (time to sleep)

**Comment by jeffwa** [comment]
@mikeerickson Sorry, forgot to post this comment last week.

Tried that, no change.

**Comment by mikeerickson** [comment]
Wow, this is a bizarre issue. I am pretty stumped?

**Comment by jeffwa** [comment]
Same here. I'm going to have to table trying to find the cause of this issue for now. Other projects at work need my attention. I wouldn't worry about it. When I get a chance to look into this again I'll let you know. Thanks again for looking into this with me.

**Comment by mikeerickson** [comment]
No worries, I completely understand.  I am beyond buried at work as well. Use this project daily and all works fine, gotta be something obscure.  Thx for efforts.

**Comment by jeffwa** [comment]
@mikeerickson, I just now got back around to dealing with this issue again. It's still there for me, this time in a different project (though it's still there in the original project.) However, I was still running 0.11.2 Yeah, I know that's *extremely* out of date) in both the of the offending projects. `npm update` wouldn't update the module to anything newer so I checked GitHub to see the current version. Seeing my installed version was out of date, I removed it and installed again and wouldn't you know it, it works! Whatever it was about my setup and the 0.11.* series is baffling, but with the latest everything works great. Thanks for all your help trying to debug this last year. I truly appreciate it.
